### PR TITLE
#204 - Long running controllers

### PR DESCRIPTION
**Pull request by mensfeld**
Now we build a controller instance per batch, this disallows us from doing some cross-batch stuff like aggregating up until we have what we need to flush to DB and so on.

This PR aims to fix that and provide a single controller instance per topic partition. It also lowers the GC usage and makes things a bit faster.

Also, the topic ownership over controller is moved to a class level.

This PR contains also refactor that extracts topic configuration controller specific behaviors into modules that are included on demand only when needed.

### CODE REVIEW

**Comment by mensfeld** [comment]
Added the #207 to this PR as well.

File: lib/karafka/base_controller.rb [hunk 1]
```diff
@@ -95,33 +98,12 @@ def params_batch=(messages)
...
-        else
-          raise Errors::InvalidProcessingBackend, topic.processing_backend
-        end
-      end
-    end
-
-    # @note We want to leave the #perform method as a public API, but just in case we will do some
-    #   pre or post processing we use call method instead of directly executing #perform
+    # will call process method of a proper backend
     def call
```
Comment (mach-kernel) [inline lib/karafka/base_controller.rb]: Much cleaner interface here

Comment (mensfeld) [reply lib/karafka/base_controller.rb]: You know, that I did this mostly for your #147 ;) to simplify things (no rush with that one though)

File: lib/karafka/base_worker.rb [hunk 2]
```diff
@@ -10,17 +10,18 @@ class BaseWorker
...
     # @return [Karafka::Controller] descendant of Karafka::BaseController that matches the topic
     #   with params_batch assigned already (controller is ready to use)
     def controller(topic_id, params_batch)
-      @controller ||= Karafka::Routing::Router.build(topic_id).tap do |ctrl|
-        ctrl.params_batch = ctrl.topic.interchanger.parse(params_batch)
-      end
+      topic = Karafka::Routing::Router.find(topic_id)
+      controller = topic.controller.new
+      controller.params_batch = topic.interchanger.parse(params_batch)
+      controller
```
Comment (mach-kernel) [inline lib/karafka/base_worker.rb]: You don't have to make this change, but I always cannot resist some playful code golf :joy::

```ruby
topic.controller.new.tap { |c| c.params_batch = topic.interchanger.parse(params_batch) }
```

File: lib/karafka/controllers/sidekiq_backend.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module Karafka
+  module Controllers
+    # Sidekiq backend that schedules stuff to Sidekiq worker for delayed execution
+    module SidekiqBackend
```
Comment (mach-kernel) [inline lib/karafka/controllers/sidekiq_backend.rb]: 👍

File: lib/karafka/controllers/includer.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  # Additional functionalities for controllers
+  module Controllers
+    # Module used to inject functionalities into a given controller class, based on the controller
+    # topic and its settings
+    # We don't need all the behaviors in all the cases, so it is totally not worth having
+    # everything in all the cases all the time
+    module Includer
+      # @param controller_class [Class] controller class, that will get some functionalities
+      #   based on the topic under which it operates
+      def self.call(controller_class)
```
Comment (mach-kernel) [inline lib/karafka/controllers/includer.rb]: This can also be a great way for other libraries to build prefab stuff into controllers.

Comment (mensfeld) [reply lib/karafka/controllers/includer.rb]: Yeah, we can come up with something more flexible in the future, but I agree, that this is a good start

**Pull request merged**